src/architecture/serialization.md

```diff
@@ -1,533 +1,201 @@
+    pb
+}
 
-        struct MessageTwoIntegers {
-            first: u64,
-            second: u64,
-        }
-    }
+fn serialize(pb: &StructName) -> Vec<u8> {
+    pb.write_to_bytes().unwrap()
 }
 ```
 
-Here the message body is serialized as a `struct` with fields `first` and `second`
-having type `u64`.
+For example, the generated `Wallet` structure, included in the
+`cryptocurrency.proto` file, which resides in the [`proto`][module] module,
+will be available using `proto::cryptocurrency::Wallet`.
 
-## Types to Be Supported in Future
+`exonum_build` also includes a function which returns the path to the `.proto`
+files that come with the `exonum` crate - `get_exonum_protobuf_files_path()`:
 
-The current version does not support the serialization of the following types,
-but it is planned to be implemented in future:
-
-- [Enums][rust_enums]
-
-## Example
-
-Consider the structure with three fields:
-
-- pub_key: `PublicKey`  
-  99ace6c721db293b0ed5b487e6d6111f22a8c55d2a1b7606b6fa6e6c29671aa1
-
-- Owner: `String`  
-  Andrew
+```rust
+let exonum_protos = get_exonum_protobuf_files_path();
+protobuf_generate(
+    "src/proto",
+    &["src/proto", &exonum_protos],
+    "example_mod.rs",
+    );
+```
 
-- Balance: `u64`  
-  1234
+After calling the `get_exonum_protobuf_files_path()` function, users can
+import protobuf descriptions from the `exonum` crate in their `.proto` files.
+For example, `import "helpers.proto";` can be used to get access to protobuf
+types `exonum.PublicKey`, `exonum.Hash`, etc. An example of such usage can be
+found in our [cryptocurrency example service][cryptocurrency].
+
+## Additional Validation for Protobuf-Generated Structures
+
+Protobuf is a versatile and flexible tool, which presents not only
+opportunities but also certain complications for the Exonum framework. For
+example, fields in protobuf cannot be fixed-size arrays, however, fixed-size
+arrays are required in Exonum (e.g. for hashes). It is possible to implement
+additional validations using the `.rs` protobuf-generated files. However, if
+users work with protobuf-generated structures, field validation would need to
+be performed every time they are used.
+
+To have validation performed only once for the whole structure, Exonum
+provides the conversion mechanism using the `ProtobufConvert` trait. This trait
+lets users automatically map
+their structures and the structures generated from `.proto` descriptions,
+providing a mechanism for validating protobuf-generated data. The structures
+for `ProtobufConvert` should have the same fields as the structures in
+`.proto` files, but can contain additional validation.
+
+The [`exonum_derive`][derive]
+crate provides the ability to use structures typical for Exonum with all
+the required validations. So when using these structures users only need to
+implement `#[derive(ProtobufConvert)]` for them. If required, users can
+implement the [`ProtobufConvert`][convert] trait for any other structure they
```

Comment:
Nit: I'd link to `docs.rs` here.

Reply:
This section of docs.rs currently fails to build, so unfortunately I cannot link to it.